alf/networks/stable_normal_projection_network.py

```diff
@@ -0,0 +1,141 @@
+
+
+@gin.configurable
+class StableNormalProjectionNetwork(NormalProjectionNetwork):
+    """Generates a tfp.distribution.Normal by predicting a mean and std.
+
+    It parameterizes the normal distributions as std=c0+1/(c1+softplus(b))
+    and mean=a*std where a and b are outputs from means_projection_layer and
+    stds_projectin_layer respectively. c0 and c1 are chosen so that
+    min_std <= std <= max_std. The advantage of this parameterization is that
+    its second order derivatives with respect to a and b are bounded even when
+    the standard deviations become very small so that the optimization is
+    stabler.
+    """
+
+    def __init__(self,
+                 sample_spec,
+                 activation_fn=None,
+                 init_means_output_factor=1e-10,
+                 mean_transform=tanh_squash_to_spec,
+                 inverse_std_transform='softplus',
+                 state_dependent_std=False,
+                 scale_distribution=False,
+                 init_std=1.0,
+                 min_std=0.0,
+                 max_std=None):
+        """Creates an instance of StableNormalProjectionNetwork.
+
+        Args:
+            sample_spec (BoundedTensorSpec): detailing the shape and dtypes of
+                samples pulled from the output distribution.
+            activation_fn (Callable): Activation function to use in dense layer.
+            init_means_output_factor (float): Output factor for initializing
+                action means weights.
+            mean_transform (Callable): Transform to apply to the calculated
+                means. Uses `tanh_squash_to_spec` by default.
+            inverse_std_transform (Callable): Currently supports tf.math.exp and
+                tf.nn.softplus. Transformation to obtain inverse std. The
+                transformed values are further transformed according to min_std
+                and max_std.
+            state_dependent_std (bool): If true, stddevs will be produced by MLP
+                from state. else, stddevs will be an independent variable.
+            scale_distribution (bool): Whether or not to use a bijector chain to
+                scale distributions to match the sample spec. Note the
+                TransformedDistribution does not support certain operations
+                required by some agents or policies such as KL divergence
+                calculations or Mode.
+            init_std (float): Initial value for standard deviation.
+            min_std (float): Minimum value for standard deviation.
+            max_std (float): Maximum value for standard deviation. If None, no
+                maximum is enforced.
+        """
+        self._min_std = min_std
+        self._max_std = max_std
+
+        c = 1 / (init_std - min_std)
+        if max_std is not None:
+            c -= 1 / (max_std - min_std)
+
+        if inverse_std_transform == 'exp':
+            std_transform = tf.math.exp
```

Comment:
numeric overflow  is very likely to happen if `tf.math.exp` was apply without clipping